# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/protocol/http/DnsAndConnectSocket.cpp
# Commit: 521c9e019678
# Full Hash: 521c9e019678f584590f63473416d40e03f33342
# Author: Valentin Gosu <valentin.gosu@gmail.com>
# Date: 2023-03-15 04:10:55
# Regressor Bug: 1805372
# File Overlap Count: 1
# Description:
#   Bug 1805372 - Pass effectiveTRRMode to nsHttpChannel from socketTransport/DnsAndConnectSocket r=necko-reviewers,kershaw
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D164539
# ==============================================================================

diff -r dfb39ca05529 -r 521c9e019678 netwerk/protocol/http/DnsAndConnectSocket.cpp
--- a/netwerk/protocol/http/DnsAndConnectSocket.cpp	Tue Mar 14 15:29:49 2023 +0200
+++ b/netwerk/protocol/http/DnsAndConnectSocket.cpp	Tue Mar 14 12:59:40 2023 +0000
@@ -420,6 +420,16 @@
   LOG(("DnsAndConnectSocket::OnLookupComplete: this=%p status %" PRIx32 ".",
        this, static_cast<uint32_t>(status)));
 
+  if (nsCOMPtr<nsIDNSAddrRecord> addrRecord = do_QueryInterface((rec))) {
+    nsIRequest::TRRMode effectivemode = nsIRequest::TRR_DEFAULT_MODE;
+    addrRecord->GetEffectiveTRRMode(&effectivemode);
+    nsITRRSkipReason::value skipReason = nsITRRSkipReason::TRR_UNSET;
+    addrRecord->GetTrrSkipReason(&skipReason);
+    if (mTransaction) {
+      mTransaction->SetTRRInfo(effectivemode, skipReason);
+    }
+  }
+
   MOZ_DIAGNOSTIC_ASSERT(request);
   RefPtr<DnsAndConnectSocket> deleteProtector(this);
 
@@ -1281,11 +1291,12 @@
 
   nsresult rv = NS_OK;
   do {
-    rv = dns->AsyncResolveNative(mHost, nsIDNSService::RESOLVE_TYPE_DEFAULT,
-                                 mDnsFlags, nullptr, dnsAndSock,
-                                 gSocketTransportService,
-                                 dnsAndSock->mConnInfo->GetOriginAttributes(),
-                                 getter_AddRefs(mDNSRequest));
+    rv = dns->AsyncResolveNative(
+        mHost, nsIDNSService::RESOLVE_TYPE_DEFAULT,
+        mDnsFlags | nsIDNSService::RESOLVE_WANT_RECORD_ON_ERROR, nullptr,
+        dnsAndSock, gSocketTransportService,
+        dnsAndSock->mConnInfo->GetOriginAttributes(),
+        getter_AddRefs(mDNSRequest));
   } while (NS_FAILED(rv) && ShouldRetryDNS());
 
   if (NS_FAILED(rv)) {